docs(elements): add notice about choosing a tag name in Elements guide

gkalpak opened the pull request:
This PR adds a warning in the Elements guide about using `@Component.selector` as the tag name for the registered custom element.
See also #40452 for context.

It also adds a couple of links for built-in APIs mentioned in the same guide.
(See individual commits.)

mary-poppins commented:
You can preview c954551 at https://pr40510-c954551.ngbuilds.io/.

petebacondarwin commented on aio/content/guide/elements.md:
```diff
@@ -64,14 +64,22 @@ create and destroy instances, and to detect and respond to changes.
 
-Use a JavaScript function, `customElements.define()`,  to register the configured constructor
-and its associated custom-element tag with the browser's `CustomElementRegistry`.
+Use the built-in [`customElements.define()`](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/define) function, to register the configured constructor and its associated custom-element tag with the browser's [`CustomElementRegistry`](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry).
```
Do we have a convention for line length in guide files?
There are some long and some short lines this file alone.

gkalpak replied:
AFAICT, the (unwritten?) rule is to use one line per sentence for all MD files. But not everyone is using (or enforcing) that rule, so we are inconsistent throughout the codebase. I try to apply the rule to the sections I edit.

petebacondarwin reviewed (approved):
One question.

Reviewed-for: global-docs-approvers

aikidave commented on aio/content/guide/elements.md:
```diff
@@ -64,14 +64,22 @@ create and destroy instances, and to detect and respond to changes.
 
-Use a JavaScript function, `customElements.define()`,  to register the configured constructor
-and its associated custom-element tag with the browser's `CustomElementRegistry`.
+Use the built-in [`customElements.define()`](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/define) function, to register the configured constructor and its associated custom-element tag with the browser's [`CustomElementRegistry`](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry).
```
```suggestion
Use the built-in [`customElements.define()`](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/define) function to register the configured constructor and its associated custom-element tag with the browser's [`CustomElementRegistry`](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry).
```

aikidave commented on aio/content/guide/elements.md:
```diff
@@ -64,14 +64,22 @@ create and destroy instances, and to detect and respond to changes.
+
+  Avoid using the [`@Component`](api/core/Component) [selector](api/core/Directive#selector) as the custom-element tag name.
+  This can lead to unexpected behavior, due to Angular creating two component instances for a single DOM element:
+  One regular Angular component and a second one via the custom element.
```
```suggestion
  One regular Angular component and a second one using the custom element.
```

gkalpak replied:
"Using" sounds a little confusing to me here. What does it refer to? Does Angular use the custom element to instantiate a second component? (This doesn't sound right 😕)

aikidave replied:
"Using" is generally considered to be a less formal, easier-to-understand alternative to "via." @TeriGlover, any thoughts?

gkalpak replied:
I've changed this to "using". @TeriGlover, ptal.

TeriGlover commented:
I agree; *using* is a better choice.

On Mon, Jan 25, 2021 at 9:57 AM David Shevitz <notifications@github.com>
wrote:

> *@aikidave* commented on this pull request.
> ------------------------------
>
> In aio/content/guide/elements.md
> <https://github.com/angular/angular/pull/40510#discussion_r563927772>:
>
> >  When the browser encounters the tag for the registered element, it uses the constructor to create a custom-element instance.
>
>  <div class="lightbox">
>    <img src="generated/images/guide/elements/createElement.png" alt="Transform a component to a custom element" class="left">
>  </div>
>
> +<div class="alert is-important">
> +
> +  Avoid using the [`@Component`](api/core/Component) [selector](api/core/Directive#selector) as the custom-element tag name.
> +  This can lead to unexpected behavior, due to Angular creating two component instances for a single DOM element:
> +  One regular Angular component and a second one via the custom element.
>
> "Using" is generally considered to be a less formal, easier-to-understand
> alternative to "via." @TeriGlover <https://github.com/TeriGlover>, any
> thoughts?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/40510#discussion_r563927772>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AR4DPGSZLBRNQJXIQYHLEXDS3WWIPANCNFSM4WNFVMQA>
> .
>


-- 
*--Teri Glover--*
*Technical Editor*

mary-poppins commented:
You can preview 774b66b at https://pr40510-774b66b.ngbuilds.io/.

gkalpak committed:
fixup! docs(elements): add MDN links for mentioned built-in APIs in Elements guide

gkalpak committed:
docs(elements): add MDN links for mentioned built-in APIs in Elements guide

gkalpak committed:
fixup! docs(elements): add notice about choosing a tag name in Elements guide

gkalpak committed:
docs(elements): add notice about choosing a tag name in Elements guide

This commit adds a warning in the Elements guide about using
`@Component.selector` as the tag name for the registered custom element.
See also #40452 for context.

mary-poppins commented:
You can preview 5ac2c92 at https://pr40510-5ac2c92.ngbuilds.io/.

TeriGlover reviewed (commented):
LGTM

TeriGlover commented:
Editing review completed.

On Tue, Jan 26, 2021 at 4:56 AM George Kalpakas <notifications@github.com>
wrote:

> *@gkalpak* commented on this pull request.
> ------------------------------
>
> In aio/content/guide/elements.md
> <https://github.com/angular/angular/pull/40510#discussion_r564489135>:
>
> >  When the browser encounters the tag for the registered element, it uses the constructor to create a custom-element instance.
>
>  <div class="lightbox">
>    <img src="generated/images/guide/elements/createElement.png" alt="Transform a component to a custom element" class="left">
>  </div>
>
> +<div class="alert is-important">
> +
> +  Avoid using the [`@Component`](api/core/Component) [selector](api/core/Directive#selector) as the custom-element tag name.
> +  This can lead to unexpected behavior, due to Angular creating two component instances for a single DOM element:
> +  One regular Angular component and a second one via the custom element.
>
> I've changed this to "using". @TeriGlover <https://github.com/TeriGlover>,
> ptal.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/40510#discussion_r564489135>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AR4DPGUVLREOXVVQH4ANTI3S323YTANCNFSM4WNFVMQA>
> .
>


-- 
*--Teri Glover--*
*Technical Editor*
